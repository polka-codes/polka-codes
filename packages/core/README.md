# Polka Codes Core

[![npm version](https://img.shields.io/npm/v/@polka-codes/core.svg)](https://www.npmjs.com/package/@polka-codes/core)
[![npm downloads](https://img.shields.io/npm/dm/@polka-codes/core.svg)](https://www.npmjs.com/package/@polka-codes/core)
[![License](https://img.shields.io/npm/l/@polka-codes/core.svg)](https://github.com/polkacodes/polkacodes/blob/main/LICENSE)

Core AI services and agent implementations for Polka Codes framework.

## Features

- Multiple AI provider support (Ollama, Anthropic, DeepSeek)
- Extensible agent architecture
- Tool integration system
- Type-safe API
- Logging and monitoring

## Installation

```bash
bun add @polka-codes/core
```

## Usage

### Basic Setup

```typescript
import { AgentBase } from '@polka-codes/core';

class MyAgent extends AgentBase {
  // Implement your agent logic
}

const agent = new MyAgent();
await agent.runTask('your-task');
```

### Available Services

- `AnthropicService`: Integration with Anthropic API
- `DeepSeekService`: Integration with DeepSeek API
- `OllamaService`: Integration with Ollama API

### Tools

The core package provides a tool system for extending functionality:

```typescript
import { ToolBase } from '@polka-codes/core';

class MyTool extends ToolBase {
  // Implement your tool logic
}
```

## Development

### Building

```bash
cd packages/core
bun run build
```

### Testing

```bash
bun test
```

---

*This README was generated by polka.codes*
