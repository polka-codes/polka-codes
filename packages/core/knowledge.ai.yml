files:
  "src/Agent/index.ts":
    description: "Base agent implementation and agent registry"
    api:
      classes:
        1:
          name: "AgentBase"
          methods:
            1:
              name: "run"
              params:
                1: { name: "task", type: "string" }
              returns: "Promise<void>"
  "src/Agent/MultiAgent.ts":
    description: "Multi-agent coordination system"
    api:
      classes:
        1:
          name: "MultiAgent"
          methods:
            1:
              name: "delegateTask"
              params:
                1: { name: "task", type: "string" }
                2: { name: "context", type: "object" }
              returns: "Promise<void>"
  "src/AiService/index.ts":
    description: "AI service abstraction layer"
    api:
      classes:
        1:
          name: "AiServiceBase"
          methods:
            1:
              name: "complete"
              params:
                1: { name: "prompt", type: "string" }
              returns: "Promise<string>"
  "src/tools/index.ts":
    description: "Tool definitions and registry"
    api:
      functions:
        1:
          name: "getAvailableTools"
          params: []
          returns: "Tool[]"
  "src/tools/updateKnowledge.ts":
    description: "Tool for updating knowledge.ai.yml files with smart merging capabilities"
    api:
      functions:
        1:
          name: "updateKnowledge"
          description: "Updates knowledge files with smart merging of YAML content"
          params:
            1: { name: "path", type: "string", description: "Directory path where knowledge file is located" }
            2: { name: "file", type: "string", description: "Optional specific file path" }
            3: { name: "knowledge", type: "string", description: "YAML content to merge" }
          returns: "Promise<ToolResponse>"

invariants:
  1: "All agents must extend AgentBase"
  2: "AI services must implement AiServiceBase"
  3: "Tools must follow the Tool interface"

patterns:
  1: "Agent implementations use prompt templates"
  2: "AI services follow a provider pattern"
  3: "Tools use XML-style formatting for execution"

rules:
  1: "Each agent must have corresponding prompt templates"
  2: "All tools must have tests"
  3: "AI service calls must be metered"
  4: "Knowledge management policies must be followed"
  5: "Use numbered dictionary format instead of arrays in knowledge files"
