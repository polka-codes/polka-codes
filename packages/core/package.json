{
  "name": "@polka-codes/core",
  "version": "0.8.17",
  "license": "AGPL-3.0",
  "author": "github@polka.codes",
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    }
  },
  "files": ["dist"],
  "scripts": {
    "build": "tsup src/index.ts --experimental-dts --format esm --clean"
  },
  "dependencies": {
    "@anthropic-ai/sdk": "0.39.0",
    "lodash": "^4.17.21",
    "openai": "^4.80.0",
    "yaml": "^2.7.1",
    "zod": "^3.24.2"
  }
}
