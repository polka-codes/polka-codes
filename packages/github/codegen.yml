schema: ./schema.docs.graphql
documents: "src/**/*.(graphql|gql)"
generates:
  src/types/github-types.ts:
    plugins:
      - "typescript"
      - "typescript-operations"
