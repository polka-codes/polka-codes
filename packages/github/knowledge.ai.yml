files:
  - path: "src/github.ts"
    description: "GitHub API client implementation"
    api:
      classes:
        - name: "GitHubClient"
          methods:
            - name: "getIssue"
              params: [{ name: "options", type: "GetIssueOptions" }]
              returns: "Promise<GitHubIssue>"
            - name: "getPullRequest"
              params: [{ name: "options", type: "GetPROptions" }]
              returns: "Promise<GitHubPR>"
  - path: "src/processBody.ts"
    description: "GitHub issue/PR body processing utilities"
    api:
      functions:
        - name: "processBody"
          params: [{ name: "body", type: "string" }]
          returns: "ProcessedBody"

invariants:
  - "All GitHub API calls must be authenticated"
  - "GraphQL queries must be defined in .gql files"

patterns:
  - "API responses are typed and validated"
  - "Error handling includes GitHub rate limit awareness"

rules:
  - "All GraphQL queries must have corresponding types"
  - "API calls must handle rate limiting"
  - "Sensitive data must not be logged"
