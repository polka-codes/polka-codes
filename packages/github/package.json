{
  "name": "@polka-codes/github",
  "version": "0.9.30",
  "license": "AGPL-3.0",
  "author": "github@polka.codes",
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    }
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsup src/index.ts --experimental-dts --format esm --clean --loader .gql=text"
  },
  "dependencies": {
    "@octokit/core": "^6.1.4",
    "mdast": "^3.0.0",
    "remark-gfm": "^4.0.1",
    "remark-parse": "^11.0.0",
    "remark-stringify": "^11.0.0",
    "unified": "^11.0.5",
    "unist-util-visit": "^5.0.0"
  }
}
