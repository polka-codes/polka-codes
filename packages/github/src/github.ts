import type { Octokit } from '@octokit/core'
import { processBody } from './processBody'
import fetchIssueQuery from './queries/fetchIssue.gql' with { type: 'text' }
import fetchPRQuery from './queries/fetchPR.gql' with { type: 'text' }
import type { FetchIssueQuery, FetchPrQuery } from './types/github-types'

const query = (octokit: Octokit) => octokit.graphql

export async function fetchIssue(owner: string, repo: string, issueNumber: number, octokit: Octokit) {
  const resp = await query(octokit)<FetchIssueQuery>(fetchIssueQuery, { owner, repo, issueNumber })
  const issue = resp.repository?.issue
  if (!issue) {
    throw new Error('Issue not found')
  }
  const issueBody = await processBody(issue.body)
  let text = `#${issue.number}: ${issue.title}
${issueBody}
`
  const comments = issue.comments?.nodes

  if (comments) {
    const totalCount = issue.comments?.totalCount ?? 0
    const skipped = totalCount - (comments?.length ?? 0)

    text += '\n============ Comments ============\n'

    if (skipped > 0) {
      text += `${skipped} comment${skipped > 1 ? 's' : ''} skipped\n`
    }

    for (const comment of comments) {
      if (!comment) {
        continue
      }
      const commentBody = await processBody(comment.body)

      const author = comment.author?.login ?? 'unknown'

      text += `${comment.createdAt} @${author}:
${commentBody}
========================
`
    }
  }
  return text
}

export async function fetchPR(owner: string, repo: string, prNumber: number, octokit: Octokit) {
  const resp = await query(octokit)<FetchPrQuery>(fetchPRQuery, { owner, repo, prNumber })
  const pr = resp.repository?.pullRequest
  if (!pr) {
    throw new Error('PR not found')
  }

  const diff = await octokit.request({
    method: 'GET',
    url: `https://api.github.com/repos/${owner}/${repo}/pulls/${prNumber}`,
    headers: {
      accept: 'application/vnd.github.v3.diff',
    },
  })

  const prBody = await processBody(pr.body)

  let text = `#${pr.number}: ${pr.title}
${prBody}
============ Diff ============
${diff.data}
`

  const comments = pr.comments?.nodes

  if (comments && comments.length > 0) {
    const totalCount = pr.comments?.totalCount ?? 0
    const skipped = totalCount - (comments?.length ?? 0)

    text += '\n============ Comments ============\n'

    if (skipped > 0) {
      text += `${skipped} comment${skipped > 1 ? 's' : ''} skipped\n`
    }

    for (const comment of comments) {
      if (!comment) {
        continue
      }
      if (comment.minimizedReason) {
        continue
      }
      const commentBody = await processBody(comment.body)

      const author = comment.author?.login ?? 'unknown'

      text += `${comment.createdAt} @${author}:
${commentBody}
========================
`
    }
  }

  const reviews = pr.reviews?.nodes

  if (reviews && reviews.length > 0) {
    const totalCount = pr.reviews?.totalCount ?? 0
    const skipped = totalCount - (reviews?.length ?? 0)

    if (skipped > 0) {
      text += `${skipped} review${skipped > 1 ? 's' : ''} skipped\n`
    }

    text += '\n============ Reviews ============\n'
    for (const review of reviews) {
      if (!review) {
        continue
      }

      const author = review.author?.login ?? 'unknown'
      const reviewBody = await processBody(review.body)

      text += `${review.createdAt} @${author}:
${reviewBody}
`
      const reviewComments = review.comments?.nodes
      if (reviewComments && reviewComments.length > 0) {
        text += '\n------------ Review Comments ------------\n'

        let lastDiff: string | undefined

        for (const comment of reviewComments) {
          if (!comment) {
            continue
          }
          if (comment.minimizedReason || comment.outdated) {
            continue
          }
          const commentBody = await processBody(comment.body)

          const author = comment.author?.login ?? 'unknown'

          if (comment.diffHunk === lastDiff) {
            text += `${comment.createdAt} @${author}:
${commentBody}
-----------------------
            `
          } else {
            text += `${comment.createdAt} @${author}:
Diff:
${comment.diffHunk}
Comment:
${commentBody}
-----------------------
`
          }
        }
      }
    }

    text += '========================\n'
  }

  return text
}
