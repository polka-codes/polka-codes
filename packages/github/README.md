# Polka Codes GitHub

[![npm version](https://img.shields.io/npm/v/@polka-codes/github.svg)](https://www.npmjs.com/package/@polka-codes/github)
[![npm downloads](https://img.shields.io/npm/dm/@polka-codes/github.svg)](https://www.npmjs.com/package/@polka-codes/github)
[![License](https://img.shields.io/npm/l/@polka-codes/github.svg)](https://github.com/polkacodes/polkacodes/blob/main/LICENSE)

GitHub API integration module for Polka Codes framework.

## Features

- Fetch GitHub issues with their details and comments
- Fetch GitHub pull requests with their details, diffs, comments, and reviews
- Process GitHub markdown content by removing HTML comments and Polka Codes specific markers
- GraphQL integration with GitHub API
- Markdown processing and transformation

## Installation

```bash
bun add @polka-codes/github
```

## Usage

### Fetching an Issue

```typescript
import { fetchIssue } from '@polka-codes/github';
import { Octokit } from '@octokit/core';

const octokit = new Octokit({ auth: 'your-github-token' });

const issueText = await fetchIssue({
  octokit,
  owner: 'owner-name',
  repo: 'repo-name',
  issueNumber: 123
});

console.log(issueText);
```

### Fetching a Pull Request

```typescript
import { fetchPR } from '@polka-codes/github';
import { Octokit } from '@octokit/core';

const octokit = new Octokit({ auth: 'your-github-token' });

const prText = await fetchPR({
  octokit,
  owner: 'owner-name',
  repo: 'repo-name',
  prNumber: 456
});

console.log(prText);
```

## Development

### Building

```bash
bun --filter @polka-codes/github build
```

### Testing

```bash
bun test
```

---

*This README was generated by polka.codes*
