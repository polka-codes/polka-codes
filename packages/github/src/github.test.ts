import { describe, expect, test } from 'bun:test'
import { Octokit } from '@octokit/core'
import { fetchIssue, fetchPR } from './github'

describe.skipIf(!process.env.GITHUB_TOKEN)('github', () => {
  test('fetchIssue', async () => {
    const octokit = new Octokit({
      auth: process.env.GITHUB_TOKEN,
    })
    const issue = await fetchIssue({ octokit, owner: 'polka-codes', repo: 'action', issueNumber: 1 })
    expect(issue).toMatchSnapshot()
  })

  test('fetchPR', async () => {
    const octokit = new Octokit({
      auth: process.env.GITHUB_TOKEN,
    })
    const pr = await fetchPR({ octokit, owner: 'polka-codes', repo: 'polka-codes', prNumber: 95 })
    expect(pr).toMatchSnapshot()
  })
})
