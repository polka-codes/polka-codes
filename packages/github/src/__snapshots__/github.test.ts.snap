// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`github fetchIssue 1`] = `
"#1: Test issue
some body

> quote

\`\`\`ts
const someCode = 'test'
\`\`\`


============ Comments ============
2025-02-11T00:42:46Z @xlc:
this is a comment

========================
2025-02-11T00:43:02Z @xlc:
another reply

========================
"
`;

exports[`github fetchPR 1`] = `
"#95: Refactor CLI Init Command and Simplify Test Cases
**Summary of Changes**:

* Refactored the CLI \`init\` command to handle base directory configuration
* Updated project config generation logic
* Simplified test cases for \`generateProjectConfig\` tool
* Removed redundant whitespace trimming in output parsing

**Highlights of Changed Code**:

* Added base directory handling in \`init.ts\` with proper directory change logging
* Updated \`generateProjectConfig.ts\` to remove redundant output trimming
* Simplified test cases in \`generateProjectConfig.test.ts\` by removing redundant whitespace test
* Modified \`makeAgentTool\` return type in \`index.ts\` to simplify tool execution

**Additional Information**:

* The changes improve maintainability and reduce unnecessary complexity in the codebase
* Test coverage remains comprehensive while being more focused on essential functionality

============ Diff ============
diff --git a/packages/cli/src/commands/init.ts b/packages/cli/src/commands/init.ts
index 982ddea..dd2a71a 100644
--- a/packages/cli/src/commands/init.ts
+++ b/packages/cli/src/commands/init.ts
@@ -25,6 +25,11 @@ export const initCommand = new Command('init')
 
 initCommand.action(async (options, command: Command) => {
   const cmdOptions = command.parent?.opts() ?? {}
+
+  parseOptions(cmdOptions) // process --base-dir
+
+  cmdOptions.baseDir = undefined // so it won't be processed again later
+
   const globalConfigPath = getGlobalConfigPath()
 
   let gloabl = options.global
@@ -76,7 +81,7 @@ initCommand.action(async (options, command: Command) => {
       }
     }
 
-    const { config, providerConfig, verbose, maxMessageCount, budget } = parseOptions(cmdOptions)
+    const { providerConfig, verbose, maxMessageCount, budget } = parseOptions(cmdOptions)
     let { provider, model, apiKey } = providerConfig.getConfigForCommand('init') ?? {}
 
     // Get provider configuration
@@ -161,8 +166,7 @@ initCommand.action(async (options, command: Command) => {
       // Generate project config
       console.log('Analyzing project files...')
 
-      const { response } = await generateProjectConfig(runner.multiAgent, undefined)
-      console.log(response)
+      const response = await generateProjectConfig(runner.multiAgent, undefined)
       generatedConfig = response ? parse(response) : {}
     }
 
diff --git a/packages/cli/src/options.ts b/packages/cli/src/options.ts
index 08f6699..e5c59d0 100644
--- a/packages/cli/src/options.ts
+++ b/packages/cli/src/options.ts
@@ -36,6 +36,7 @@ export function parseOptions(options: CliOptions, cwdArg?: string, home: string
   if (options.baseDir) {
     process.chdir(options.baseDir)
     cwd = options.baseDir
+    console.log('Changed working directory to', cwd)
   } else {
     cwd = process.cwd()
   }
diff --git a/packages/core/src/AiTool/createNewProject.ts b/packages/core/src/AiTool/createNewProject.ts
index 1f283e8..0a14026 100644
--- a/packages/core/src/AiTool/createNewProject.ts
+++ b/packages/core/src/AiTool/createNewProject.ts
@@ -30,7 +30,7 @@ const prompt = \`You are an AiTool designed to assist users in creating new proje
    - Based on the collected information, generate a .polkacodes.yml configuration file that includes:
      - scripts section with common development commands (test, format, check, etc.)
      - rules section reflecting project conventions and tools
-     - excludeFiles section for sensitive and generated files
+     - excludeFiles section for sensitive files only
    - Example structure:
      \\\`\\\`\\\`yaml
      scripts:
diff --git a/packages/core/src/AiTool/generateProjectConfig.test.ts b/packages/core/src/AiTool/generateProjectConfig.test.ts
index a1d70b2..41013ff 100644
--- a/packages/core/src/AiTool/generateProjectConfig.test.ts
+++ b/packages/core/src/AiTool/generateProjectConfig.test.ts
@@ -26,25 +26,6 @@ excludeFiles:
     expect(result).toBe(output)
   })
 
-  test('should trim whitespace in output', () => {
-    const output = \`
-scripts:
-  test:
-    command: "bun test"
-    description: "Run tests"
-
-rules:
-  - "Use bun as package manager"
-
-excludeFiles:
-  - ".env"
-  - "dist/"
-    \`
-
-    const result = generateProjectConfig.parseOutput(output)
-    expect(result).toBe(output.trim())
-  })
-
   test('should handle output with empty excludeFiles section', () => {
     const output = \`scripts:
   test:
diff --git a/packages/core/src/AiTool/generateProjectConfig.ts b/packages/core/src/AiTool/generateProjectConfig.ts
index c3ed668..a7cf9cc 100644
--- a/packages/core/src/AiTool/generateProjectConfig.ts
+++ b/packages/core/src/AiTool/generateProjectConfig.ts
@@ -73,7 +73,7 @@ export default {
     return ''
   },
   parseOutput: (output: string) => {
-    return output.trim()
+    return output
   },
   agent: 'analyzer',
 } as const satisfies AiToolDefinition<void>
diff --git a/packages/core/src/AiTool/index.ts b/packages/core/src/AiTool/index.ts
index 772bcc2..031b0ec 100644
--- a/packages/core/src/AiTool/index.ts
+++ b/packages/core/src/AiTool/index.ts
@@ -51,7 +51,7 @@ export const makeTool = <T extends AiToolDefinition<any, any>>(definition: T) =>
 }
 
 export const makeAgentTool = <T extends AiToolDefinition<any, any>>(definition: T) => {
-  return async (agent: MultiAgent, params: GetInput<T>): Promise<{ response: GetOutput<T>; usage: ApiUsage }> => {
+  return async (agent: MultiAgent, params: GetInput<T>): Promise<GetOutput<T>> => {
     return executeAgentTool(definition, agent, params)
   }
 }


============ Comments ============
2025-02-11T01:51:40Z @xlc:
test comment

========================

============ Reviews ============
2025-02-11T01:51:51Z @xlc:
review


------------ Review Comments ------------
2025-02-11T01:51:51Z @xlc:
Diff:
@@ -76,7 +81,7 @@ initCommand.action(async (options, command: Command) => {
       }
     }
 
-    const { config, providerConfig, verbose, maxMessageCount, budget } = parseOptions(cmdOptions)
+    const { providerConfig, verbose, maxMessageCount, budget } = parseOptions(cmdOptions)
Comment:
test review

-----------------------
2025-02-11T01:52:15Z @xlc:
Diff:
@@ -30,7 +30,7 @@ const prompt = \`You are an AiTool designed to assist users in creating new proje
    - Based on the collected information, generate a .polkacodes.yml configuration file that includes:
      - scripts section with common development commands (test, format, check, etc.)
      - rules section reflecting project conventions and tools
Comment:
some comment

-----------------------
2025-02-11T01:52:21Z @xlc:
Diff:
@@ -30,7 +30,7 @@ const prompt = \`You are an AiTool designed to assist users in creating new proje
    - Based on the collected information, generate a .polkacodes.yml configuration file that includes:
      - scripts section with common development commands (test, format, check, etc.)
      - rules section reflecting project conventions and tools
Comment:
some reply

-----------------------
========================
"
`;
